mm2src/coins/z_coin/storage.rs

```diff
@@ -1,5 +1,215 @@
+/// `scan_cached_blocks` and passing `None` for the optional `limit` value.
+///
+/// This function pays attention only to cached blocks with heights greater than the
+/// highest scanned block in `data`. Cached blocks with lower heights are not verified
+/// against previously-scanned blocks. In particular, this function **assumes** that the
+/// caller is handling rollbacks.
+///
+/// For brand-new light client databases, this function starts scanning from the Sapling
+/// activation height. This height can be fast-forwarded to a more recent block by
+/// initializing the client database with a starting block (for example, calling
+/// `init_blocks_table` before this function if using `zcash_client_sqlite`).
+///
+/// Scanned blocks are required to be height-sequential. If a block is missing from the
+/// cache, an error will be returned with kind [`ChainInvalid::BlockHeightDiscontinuity`].
+///
+#[cfg(not(target_arch = "wasm32"))]
+pub async fn scan_cached_block(
+    data: DataConnStmtCacheWrapper,
+    params: &ZcoinConsensusParams,
+    block: &CompactBlock,
+    last_height: &mut BlockHeight,
+) -> Result<(), ValidateBlocksError> {
+    let mut data_guard = data.inner();
+    // Fetch the ExtendedFullViewingKeys we are tracking
+    let extfvks = data_guard.get_extended_full_viewing_keys().await?;
+    let extfvks: Vec<(&AccountId, &ExtendedFullViewingKey)> = extfvks.iter().collect();
+
+    // Get the most recent CommitmentTree
+    let mut tree = data_guard
+        .get_commitment_tree(*last_height)
+        .await
+        .map(|t| t.unwrap_or_else(CommitmentTree::empty))?;
+
+    // Get most recent incremental witnesses for the notes we are tracking
+    let mut witnesses = data_guard.get_witnesses(*last_height).await?;
+
+    // Get the nullifiers for the notes we are tracking
+    let mut nullifiers = data_guard.get_nullifiers().await?;
+
+    let current_height = block.height();
+    // Scanned blocks MUST be height-sequential.
+    if current_height != (*last_height + 1) {
+        return Err(ValidateBlocksError::block_height_discontinuity(
+            *last_height + 1,
+            current_height,
+        ));
+    }
+
+    let block_hash = BlockHash::from_slice(&block.hash);
+    let block_time = block.time;
+
+    let txs: Vec<WalletTx<Nullifier>> = {
+        let mut witness_refs: Vec<_> = witnesses.iter_mut().map(|w| &mut w.1).collect();
+        scan_block(
+            params,
+            block.clone(),
+            &extfvks,
+            &nullifiers,
+            &mut tree,
+            &mut witness_refs[..],
+        )
+    };
+
+    // Enforce that all roots match. This is slow, so only include in debug builds.
+    #[cfg(debug_assertions)]
+    {
+        let cur_root = tree.root();
+        for row in &witnesses {
+            if row.1.root() != cur_root {
+                return Err(Error::InvalidWitnessAnchor(row.0, current_height).into());
+            }
+        }
```

Comment:
much simpler I believe with less nested block

```suggestion
        if witnesses.iter().any(|row| row.1.root() != cur_root) {
            return Err(Error::InvalidWitnessAnchor(row.0, current_height).into());
        }
```

Reply:
thanks, done